### PR TITLE
android: support api level less than 21

### PR DESCRIPTION
**Pull request by kkdaemon**
after introducing uv_os_homedir minimum android api level is 21 due to getpwuid_r function use
I'd like to use libuv with android api less than 21, e.g 14
getpwuid_r can be accessed through dlsym, I tested it on android 4.0.4, 4.1.2, 4.2.1, arm and x86

also I saw issue https://github.com/libuv/libuv/issues/406

### CODE REVIEW

File: src/unix/core.c [hunk 1]
```diff
@@ -1018,6 +1021,10 @@ int uv_os_homedir(char* buffer, size_t* size) {
   size_t len;
   long initsize;
   int r;
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 21
+  typedef int(*fptr_getpwuid_r)(uid_t, struct passwd*, char*, size_t, struct passwd**);
```
Comment (bnoordhuis) [inline src/unix/core.c]: This line is too long, it should be <= 80 columns.  What you could do is:
```c
int (*getpwuid_r)(uid_t, struct passwd*, char*, size_t, struct passwd**);
// further down below
getpwuid_r = dlsym(RTLD_DEFAULT, "getpwuid_r");
```
That way, you don't need the third `#if`, you can just call `getpwuid_r(...)`.

File: src/unix/core.c [hunk 2]
```diff
@@ -1039,6 +1046,12 @@ int uv_os_homedir(char* buffer, size_t* size) {
     return 0;
   }
 
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 21
+  fn_getpwuid_r = (fptr_getpwuid_r)dlsym(RTLD_DEFAULT, "getpwuid_r");
+  if (!fn_getpwuid_r)
```
Comment (bnoordhuis) [inline src/unix/core.c]: Can you use an explicit `== NULL` here?

**Comment by bnoordhuis** [comment]
Can you explain in the commit log what you changed and why?

**Comment by kkdaemon** [comment]
I have added commit descriptive message and prettified code

File: src/unix/core.c [hunk 3]
```diff
@@ -1008,6 +1008,9 @@ int uv__dup2_cloexec(int oldfd, int newfd) {
   }
 }
 
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 21
+#  include <dlfcn.h>  /* for dlsym */
+#endif
```
Comment (saghul) [inline src/unix/core.c]: Maybe move this to the top with the other includes?

**Comment by saghul** [comment]
One nit, otherwise LGTM.

File: src/unix/core.c (hunk 3 above)
Comment (kkdaemon) [inline src/unix/core.c]: ok

**Comment by bnoordhuis** [comment]
No way to test it but LGTM.

**Comment by kkdaemon** [comment]
I tested it using following devices
android 4.1.2 zte nubia z5 and noname chinese tablet
android 4.2.1 lenovo k900 x86
android 4.0.4 samsung galaxy s3 i9300
android 5.0.2 samsung galaxy tab 4

libuv was built against android api 14 (make-standalone-toolchain.sh --platform=android-14)

**Comment by saghul** [comment]
@kkdaemon don't worry, we trust you :-) Landed in 43994b4, thanks a lot!

**Pull request closed without merging**